lib.teams: Sync from GitHub (members/shortName/scope)

infinisil opened the pull request:
This implements the basis of https://github.com/NixOS/nixpkgs/issues/447514#issuecomment-3366610483 by:
- Adding a workflow that runs weekly to synchronise all teams with Nixpkgs access to the `maintainers/github-teams.json` file
- Populating the `.members`, `.shortName` and `.description` of `lib.teams` that have `.github` set based on `github-teams.json`
- Adding an automatically populated `.githubMaintainer` field containing contact information of the GitHub team maintainers, making it easier to know who to ask to join such a team

In order to bring the state of `lib.teams` in sync with GitHub, I compared the outputs of the two and made two commits to update either `lib.teams` or `github-teams.json`, see commit messages for details.

## TODO

- [x] Test the workflow: https://github.com/infinisil-test-org/nixpkgs/pull/47
- [x] Make all the changes listed in the "maintainers/github-teams.json: Manual adjustment of necessary changes" commit subject
- [x] Make sure `lib/tests/teams.nix` works
- [x] Code owners (especially for `maintainers/github-teams.json`, somebody needs to merge the updates)
- [x] Docs
- [x] Check performance
- [x] Test that CI fails when a GitHub user part of a team doesn't have a maintainers entry 

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

infinisil committed:
lib.teams: Some changes based on current GitHub state

Including:
- Removing a `.github` for a team that doesn't exist
- Adding people to `lib.teams` that are part of the GitHub team but were never
  part of the team in `lib.teams`
- Adjusting capitalisation of a linked GitHub team

While there's more differences, all others should be done on the GitHub
side, because it's things like adding/removing team members that have
added/removed themselves to/from `lib.teams` in the past.

infinisil committed:
workflows/team-sync: init

Creates a team sync workflow that pushes the current state of teams to a
JSON file, which can then be ingested by `lib.teams` to expose member
lists.

Co-Authored-By: Alexander Bantyev <alexander.bantyev@tweag.io>

infinisil committed:
maintainers/github-teams.json: Automated sync

infinisil committed:
maintainers/github-teams.json: Manual adjustment of necessary changes

This makes `lib.teams` be in sync with the GitHub state.

Add the marketing-team so that all teams in `lib.teams` can be requested for reviews.

Add these members to the respective teams:
- beam: minijackson (53e70ab75490df87a165d704a9e98312b3f5137b)
- beam: savtrip (f9ff37cdc3ade4bc45f501af0e7f7520a9543f1c)
- cuda: prusnak (748896a2db6cafb61e6b29dfa71f904293443f6b)
- geospatial: autra (7de303b4518766db2ddca9dfcbdc823430fe3869)
- geospatial: l0b0 (510eb5118acff1546bb0037c6f3168496675e1e8)
- gnome: bobby285271 (066634e509988232b435cf2e64c778ec4508a4b4)
- gnome: dasj19 (c57c9362815e48681e9a52ec86cf81e38fe1ce6c)
- k3s: heywoodlh (2ca18a8b175c961eef8aa3637c490de833fcccef)
- k3s: rorosen (b9f397d445b52316eaf08b9d5072fe194ab482ec)
- php: piotrkwiecinski (148e322192b7a451344fe335cc38b95043ee37d4)
- qt-kde: mjm (f6bb8a922fef75dd6a81af0cfe25daca6e3b9c4b)
- qt-kde: NickCao (f6bb8a922fef75dd6a81af0cfe25daca6e3b9c4b)
- qt-kde: ilya-fedin (f6bb8a922fef75dd6a81af0cfe25daca6e3b9c4b)
- qt-kde: K900 (f6bb8a922fef75dd6a81af0cfe25daca6e3b9c4b)
- qt-kde: SuperSandro2000 (f6bb8a922fef75dd6a81af0cfe25daca6e3b9c4b)
- qt-kde: LunNova (f6bb8a922fef75dd6a81af0cfe25daca6e3b9c4b)

Remove these members from the respective teams:
- geospatial: das-g (b93d987a8441384a6bac336318b1582d52305412)
- haskell: expipiplus1 (e134422465913332a5c9cb3ec9161650f7d459c8)
- haskell: peti (https://github.com/NixOS/nixpkgs/pull/450864#issuecomment-3431029084)
- k3s: yajo (8b238112108d8c366856434a3e617f60cd250758)
- k3s: superherointj (7354bda2081a43cfe13f7df67b09af90b309615a)
- ocaml: superherointj (7354bda2081a43cfe13f7df67b09af90b309615a)
- php: drupol (cabc16dc599da7ca2b320fafe80e256d0dfa5b94)
- php: globin (e5d552f5b94c26e574a6333060d0864760034ee3)
- documentation-team: fricklerhandwerk (https://discourse.nixos.org/t/the-next-chapter-in-nix-documentation/68425)

Copy these team descriptions from `lib.teams.*.scope` to the GitHub team description:
- nixpkgs-ci: Maintain Nixpkgs' in-tree Continuous Integration, including GitHub Actions
- cuda: Maintain CUDA-enabled packages
- darwin: Maintain core platform support and packages for macOS and other Apple platforms
- documentation-team: Maintain nixpkgs/NixOS documentation and tools for building it. https://nixos.org/community/teams/documentation
- enlightenment: Maintain Enlightenment desktop environment and related packages
- flutter: Maintain Flutter and Dart-related packages and build tools
- geospatial: Maintain geospatial, remote sensing and OpenStreetMap software
- golang: Maintain Golang compilers
- haskell: Maintain Haskell packages and infrastructure
- k3s: Maintain K3s package, NixOS module, NixOS tests, update script
- lisp: Maintain the Lisp ecosystem
- lua: Maintain the lua ecosystem
- lumina: Maintain lumina desktop environment and related packages
- lxqt: Maintain LXQt desktop environment and related packages
- neovim: Maintain the vim and neovim text editors and related packages
- ocaml: Maintain the OCaml compiler and package set
- pantheon: Maintain Pantheon desktop environment and platform
- qt-kde: Maintain the Qt framework, KDE application suite, Plasma desktop environment and related projects
- nixos-release-managers: Manage the current nixpkgs/NixOS release
- rocm: Maintain ROCm and related packages
- rust: Maintain the Rust compiler toolchain and nixpkgs integration
- systemd: Maintain systemd for NixOS
- xen-project: Maintain the Xen Project Hypervisor and the related tooling ecosystem

infinisil committed:
lib.teams: Populate fields from synced GitHub state

The before and after of

    nix-instantiate --eval -A lib.teams --strict --json | jq 'walk(if type == "array" then sort else . end)'

has been ensured to be negligible, only consisting of minor team
shortName and scope differences

infinisil committed:
lib.teams: Add githubMaintainers field

This is so you can figure out who to ask to get added to the team

Also avoid an unecessary `{ inherit lib; }` argument

infinisil committed:
maintainers/scripts/feature-freeze-teams.pl: Use GitHub CLI

Makes it much easier and less involved to run

infinisil committed:
maintainers/scripts/feature-freeze-teams.pl: Output result to stdout

infinisil committed:
lib.teams: Remove unused teams

infinisil committed:
lib.teams: Remove dummy Nixpkgs committers team and hardcode into feature freeze script

There's no reason this needs to be exposed under `lib.teams`

infinisil commented on maintainers/team-list.nix:
```diff
@@ -399,31 +325,21 @@ with lib.maintainers;
     scope = "Dummy team for the #everyone else' section during feture freezes, not to be used as package maintainers!";
     shortName = "Everyone else";
     enableFeatureFreezePing = true;
   };
```
Done now, `syncGitHub` was removed

infinisil commented on maintainers/github-teams.json:
```diff
@@ -0,0 +1,6508 @@
+    "parent": null,
+    "permission": "admin"
+  },
+  "nixpkgs-maintainers": {
```
I now changed it so that the nixpkgs-maintainers, nixpkgs-committers and voters* teams won't be synced. We can still remove this restriction for nixpkgs-maintainers later if necessary.

infinisil commented on maintainers/github-teams.json:
I removed the voters team now.

But the further cross check can be done in future PRs, it's not necessary for this one.

wolfgangwalther reviewed (dismissed):
Blocking until https://github.com/NixOS/nixpkgs/pull/450864#discussion_r2422679485 is figured out. I will open a PR for a policy later.

wolfgangwalther commented on ci/get-teams.sh:
> I'm happy to adhere to any policy, but only once it exists, otherwise I'd just be wasting my time.

And to be clear: There is an easy way out here. You don't need to wait for the policy PR. Just write that thing in `actions/github-script` and we can move quickly. I'm ready to help you to do that! It is really not hard.

wolfgangwalther commented on maintainers/team-list.nix:
```diff
@@ -569,6 +572,7 @@ with lib.maintainers;
       maralorn
       sternenseemann
       wolfgangwalther
+      peti
```
Pretty sure peti was removed from the team on purpose. I don't think adding him would be in his interest, because that involves a lot of notifications.

I last asked about this here: https://github.com/NixOS/nixpkgs/pull/354409#discussion_r2188192960

"honorary member" reads to me as: Let's remove him from the GitHub team. He doesn't maintain these other repos anyway - and if he ever needs access, it will be easy enough to give it to him.

wolfgangwalther replied:
This was meant as "if keeping him in GitHub only creates technical problems". We want these lists to be in sync, so he should now be removed from the GItHub team. The same applies to expipiplus1, btw.

wolfgangwalther replied:
Clarified in https://github.com/NixOS/nixpkgs/pull/450864#issuecomment-3431029084.

I added peti directly to the cabal2nix and jailbreak-cabal repos. I don't have maintainer rights for the GitHub team itself, so either a maintainer or org owner can remove him there.

wolfgangwalther commented on maintainers/README.md:
```diff
@@ -158,13 +158,17 @@ If the team lists no specific membership policy, feel free to merge changes to t
 > [!TIP]
 > The team name should be as short as possible; because it is nested under the maintainers group, no -maintainers suffix is needed.
 
+After the first [weekly team sync](../.github/workflows/team-sync.yml) with the new team, it's then also possible to link it to the entry in `team-list.nix` by setting its `github` field to the GitHub team name.
```
I think we should rewrite this along the lines of the following workflow:
- Open the PR with your new team, include the `github` field immediately.
- Create the GItHub team and request maintainer group parent.
- Have a -core member or org owner approve it.
- (Optionally manually trigger a sync of the workflow - let's add a manual trigger on top of the scheduled!)
- Make sure we only merge the PR once the team is synced. Would this currently fail CI?

TLDR: We should not suggest to create teams first and then add the github field later - let's make sure this all happens immediately.

infinisil replied:
Hmm, I'm not sure about this, because indeed CI will fail if the team isn't synced yet, so in practice this means the PR author will need to close/reopen or rebase the PR to trigger a new CI run once synced, which is quite involved for something so simple.

I think your goals can probably be solved better differently:
- If the goal is to have a place to explicitly request the nesting under nixpkgs-maintainers in case the native GitHub parenting request gets lost, I recommend opening an explicit issue for that, a-la https://github.com/NixOS/nixpkgs/issues/234293. Compared to committer requests, there's no additional privileges to gain here, so there shouldn't ever be longer discussions either.
- If the goal is to move towards having all Nixpkgs teams be in `lib.teams`, this can be done without any human involvement by just more or less programmatically adding all the synced teams to `lib.teams`. So once the parenting request is accepted, it will show up in `lib.teams` automatically a few days later.

wolfgangwalther replied:
My goal is neither of these. Instead it is to merge a consistent state immediately on the first PR.

This is similar to the problems I see with maintainers - I posted a long comment on that in the issue. They are added to the repo first, then invited - this leaves a phase of inconsistency until they accept the invite. Or if they don't, it messes more things up. My goal is to eventually change the maintainer stuff, so that this is immediately consistent.

So let's try to achieve the same here: You shouldn't merge a team with an empty member list, but you also shouldn't merge a team with a temporary member list that does not reflect the already created GitHub team. By sorting everything out *before* merging the PR, we have a nice checklist in form of CI checks in the PR: CI is still failing? Yes, that's because the GitHub team was not created or synced, yet. That's clean and consistent. Once it is, we can re-run CI and can tell everything is good now.

philiptaron replied:
I'm fine with merging this async thing and working towards the immediate state you sketch out @wolfgangwalther -- the rate of team formation and team change is slow enough that we can work out the kinks of the rest of the PR without blocking the PR on getting _this_ improvement detail nailed down.

wolfgangwalther commented on ci/github-script/get-teams.js:
```diff
@@ -0,0 +1,83 @@
+            await github.paginate(github.rest.teams.listMembersInOrg, {
+              org: context.repo.owner,
+              team_slug: team.slug,
+            }),
```
I believe my comment elsewhere, to reduce duplication in the .json file, would only require this:


```suggestion
              org: context.repo.owner,
              team_slug: team.slug,
              role: 'member',
            }),
```

(ofc + relevant nix code)

wolfgangwalther reviewed (dismissed):
Putting another tiny blocking review for this comment: https://github.com/NixOS/nixpkgs/pull/450864#discussion_r2449156674

(peti should not be added to haskell team, we don't want to accidentally spam him with notifications)

MattSturgeon commented on lib/tests/teams.nix:
```diff
@@ -31,6 +31,10 @@ let
           default = "";
         };
+        githubMaintainers = lib.mkOption {
+          type = types.listOf (types.submodule (import ./maintainer-module.nix { inherit lib; }));
```
I assume this is unrelated to this PR, but why does `maintainer-module.nix` need to have `{ inherit lib; }` applied? All modules can access `lib` via module args.

MattSturgeon reviewed (commented):
I'll try and re-read the bulk of the diff again when my eyes are less tired, but it SGTM on a first pass.

Thanks for taking the time to re-write this!

sternenseemann commented:
@peti should in any case have access to the @NixOS/haskell controlled repositories since he also has the corresponding write access to Hackage (as does @expipiplus1, by the way), so it makes sense to match this information up.

We could just add him individually to the repositories instead of having him in the teams, whatever he prefers.

peti commented:
The current status quo works great for me. So if that's possible, please use the second option, e.g. no GitHub team membership but access to the repositories.Thank you very much guys!
@peti would you prefer the option with more notifications and involvement or the one with less (== status quo, just differently expressed)?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

infinisil commented on ci/github-script/get-teams.js:
```diff
@@ -0,0 +1,83 @@
+            maintainers,
+            members,
+            name: team.name,
+            parent: team.parent != null ? team.parent.slug : null,
```
`pull` is read access, there's also `triage` and `push` (write access) for other teams

infinisil replied:
Marketing team is like that because I manually added it in https://github.com/NixOS/nixpkgs/commit/c3395b32d34d5ea064a6e6fb4119e474541fba82, see the remaining TODO in the PR description

wolfgangwalther replied:
Ah, so that's a change that was manually done to the file, but will be reflected via GitHub modifications later? Got it.

Still: The field is useless in this form, I think.

wolfgangwalther commented on maintainers/computed-team-list.nix:
```diff
@@ -0,0 +1,51 @@
+    lib.groupBy (attrs: toString attrs.githubId) (
+      lib.mapAttrsToList (name: attrs: attrs) lib.maintainers
+    )
+  );
```
> Personally, I wonder if a pipe is more readable:

I find it more readable, yes.

infinisil commented on .github/workflows/team-sync.yml:
```diff
@@ -0,0 +1,77 @@
+          add-paths: maintainers/github-teams.json
+          author: ${{ steps.user.outputs.git-string }}
+          committer: ${{ steps.user.outputs.git-string }}
+          commit-message: "maintainers/github-teams.json: Automated sync"
```
> But this is _automation_

I'd agree if it was automation to check such a convention, but this is automation for something completely different.

> We are encoding the style for this as code - this is exactly what the style-based reviews thing suggests as the best solution.

Not sure where you're reading that. The established policy doesn't mention it, and https://github.com/NixOS/nixpkgs/issues/387072 just mentions "Encoding lints in CI", which this is not. Establishing a policy or CI enforcing one is not the same as an arbitrary subset (however large) of automation adhering to an unwritten policy.

> At some point we must be allowed to discuss these things.

You're free to discuss it in the issue made for this: https://github.com/NixOS/nixpkgs/issues/281573.

> Edit: But to be crystal-clear - none of these suggestions are blocking, as I have indicated by making one of them blocking explicitly.

That's not relevant. The policy I mentioned is clear about that:

> Any style choices not covered here but that can be expressed as general rules should be left at the discretion of the authors of changes and **_not_ commented in reviews**.

philiptaron commented on maintainers/team-list.nix:
```diff
@@ -1153,13 +1146,6 @@ with lib.maintainers;
   };
 
-  ruby = {
-    members = [ ];
```
67d3432a25d5f1dc8e2b4a9538923fd8629037d8 from https://github.com/NixOS/nixpkgs/pull/443679 was the last big update I see. Maybe @kirillrdy or @l0b0 want to resuscitate this team?

infinisil committed:
lib.teams: Add githubId from associated github teams

This will allow the code for https://github.com/NixOS/nixpkgs/issues/447514
to make sure that the right team is requested for review,
even if it has been renamed in the meantime.

While the matching is currently based on the teams slug/name and not the
id, renames won't cause problems with `lib.teams`, because CI would
error if there's no match. Changing this can be future work.

philiptaron commented on ci/github-script/get-teams.js:
```diff
@@ -0,0 +1,78 @@
+        users
+          .sort((user1, user2) => (user1.login > user2.login ? 1 : -1))
+          // biome-ignore lint/performance/noAccumulatingSpread: We don't care about time complexity
+          .reduce((acc, user) => ({ ...acc, [user.login]: user.id }), {})
```
To ameliorate this, you can use mutation:

```suggestion
.reduce((acc, user) => ({ acc[user.login]: user.id; return acc; }, {})
```

unformatted, but you get the idea.

philiptaron commented on ci/github-script/get-teams.js:
```diff
@@ -0,0 +1,78 @@
+  const sorted = Object.keys(result)
+    .sort()
+    // biome-ignore lint/performance/noAccumulatingSpread: We don't care about time complexity
+    .reduce((acc, key) => ({ ...acc, [key]: result[key] }), {})
```
Same thing here about using mutating to avoid the accidental quadratic nature

philiptaron commented on ci/github-script/get-teams.js:
```diff
@@ -0,0 +1,78 @@
+
+    /// Process a list of teams and append to the result variable
+    async function processTeams(teams) {
+      for (const team of teams) {
```
If this is slow (likely will not be) we can do something like the following. Note that this code is definitely not tested, just sketched out off the top of my head.

```javascript
async function processTeamsParallel(teams, result = {}) {
  if (teams.length === 0) return result;

  // Process all teams at this level in parallel
  await Promise.all(
    teams.map(async (team) => {
      core.notice(`Processing team ${team.slug}`);

      // Skip excluded teams
      if (excludeTeams.some((r) => r.test(team.slug))) return;

      // Fetch members and maintainers concurrently
      const [members, maintainers] = await Promise.all([
        github
          .paginate(github.rest.teams.listMembersInOrg, {
            org: context.repo.owner,
            team_slug: team.slug,
            role: "member",
          })
          .then(makeUserSet),
        github
          .paginate(github.rest.teams.listMembersInOrg, {
            org: context.repo.owner,
            team_slug: team.slug,
            role: "maintainer",
          })
          .then(makeUserSet),
      ]);

      result[team.slug] = {
        id: team.id,
        name: team.name,
        description: team.description,
        members,
        maintainers,
      };

      // Recurse into child teams
      const children = await github.paginate(
        github.rest.teams.listChildInOrg,
        {
          org: context.repo.owner,
          team_slug: team.slug,
        }
      );

      await processTeamsParallel(children, result);
    })
  );

  return result;
}
```

infinisil replied:
It's indeed a bit slow, 30-60s, but the code you proposed doesn't work as is, I encourage making a follow-up PR instead :)

philiptaron reviewed (approved):
I support merging this as-is and iterating.

wolfgangwalther commented on .github/workflows/team-sync.yml:
```diff
@@ -0,0 +1,76 @@
+name: Teams
```
Since we changed the name of the workflow, let me add a suggestion to change the *filename* accordingly to `teams.yml`. I wouldn't comment on it, if a later change wouldn't cause this to be a separate "workflow" in the Github Actions UI, which would result in two entries in https://github.com/NixOS/nixpkgs/actions with the same user-facing name. For that reason a later change is undesirable.

wolfgangwalther reviewed (commented):
I think we should do https://github.com/NixOS/nixpkgs/pull/450864/files#r2452738182, too - I like that suggestion.

Other than that, this looks great to me.

Thank you!

infinisil commented on ci/github-script/get-teams.js:
```diff
@@ -0,0 +1,78 @@
+          // biome-ignore lint/performance/noAccumulatingSpread: We don't care about time complexity
+          .reduce((acc, user) => ({ ...acc, [user.login]: user.id }), {})
+      )
+    }
```
This works, also applied for the other case!

infinisil commented:
Thanks!

I agree that we should do that at some point, but not in this PR, because it would blow up the scope more than necessary, and this PR is already on the upper end of reviewability and reviews. PRs should be as small in scope as possible, so that we can merge fast and make follow-up PRs.

infinisil commented:
Backport not needed because the workflow runs on a schedule, which only trigger on the default branch.

We could consider whether the team sync and the weekly PRs should all get backported though.

wolfgangwalther commented:
> We could consider whether the team sync and the weekly PRs should all get backported though.

And yes, I think we should do that, too. That's because requesting reviewers always works with the target branch's data, so we want up2date data on the stable branches, too. That means we need to backport the sync, which in turn means we need to backport everything leading up to it.

@infinisil are you working on the backport or should i?

wolfgangwalther commented:
Backport in #456422.

mdaniels5757 commented:
This makes CI fail for me in #455942, because the maintainer I am proposing to drop is a member of a team, and I remove their maintainers entry.

This makes sense, and the error message was even descriptive of the actual problem! But I'm not really sure what the solution is. Should I ping an org owner, or someone listed in the "maintainers" field of maintainers/github-teams.json? When should I do so: now (during the one-week waiting period), or later? Would they be allowed to remove membership before the waiting period is up? And will we need to wait (in the worst case) a second week for the teams to sync, or is manually running the sync workflow considered to be a reasonable option?

Ideally, I think the answers to these would be addressed in the error messages? I only knew to "blame" this because I had happened to see this PR (although I'm sure grepping for the error message would have gotten me here too), and I only noticed the significance of the "maintainers" field on the second read of this PR.

wolfgangwalther commented:
Another option is also to remove the maintainer by hand from the synced file, so that CI in the "drop the maintainer" PR is happy. Once this is merged, the weekly sync will try to bring back the maintainer, at which point the org owner can clean this up.

(this is a bit annoying, because effectively only org owners can merge the weekly sync that way...)

We might have just hit a serious limitation of treating GitHub as the source of truth...
